### DISCUSSION TITLE
Nonlinear dependence on control parameter

### DISCUSSION THREAD

**Discussion by will-atom** [JuliaQuantumControl/QuantumControl.jl#42] [Q&A] [answered]
Is it possible to define a nonlinear dependence on a control parameter? 

For example, I might create a Hamiltonian function 
`hamiltonian(H0, (H1,\epsilon))`

But I would like to have another contribution `(H2,\epsilon^2)`. Of course I know that this is not possible exactly as I've written it here, because `\epsilon` is a function of time. 

What would be the best way to implement this dependence?

**Comment by goerz** [comment, marked as answer]
Yes, it should be pretty straightforward in principle:

```
julia> using LinearAlgebra

julia> using QuantumControl: hamiltonian

julia> using QuantumControlTestUtils.RandomObjects: random_matrix

julia> using QuantumControl.PulseParametrizations: SquareParametrization, ParametrizedAmplitude

julia> using QuantumControl.Controls: get_controls, evaluate

julia> N = 10;

julia> H0 = random_matrix(N; hermitian=true);

julia> H1 = random_matrix(N; hermitian=true);

julia> H2 = random_matrix(N; hermitian=true);

julia> ϵ(t) = 0.5;

julia> a = ParametrizedAmplitude(ϵ; parametrization=SquareParametrization());

julia> H = hamiltonian(H0, (H1, ϵ), (H2, a));

julia> get_controls(H)
(ϵ,)

julia> norm(Array(evaluate(H, 0.0)) - (H0 + 0.5 * H1 + 0.5^2 * H2))
0.0
```

Except that there was a bug https://github.com/JuliaQuantumControl/QuantumControl.jl/issues/43, which I just fixed.

Can you work with the current `master` branch of the JuliaQuantumControl packages? Otherwise, I have make a new release (there's been a few accumulated changes). I could probably do that tomorrow.

**Reply by will-atom** [reply]
I'm not in a huge rush, thanks for this! Happy to use it on master whenever you release the changes.

**Reply by will-atom** [reply]
Is it possible to have `ϵ` also be a `ParametrizedAmplitude` in this situation? For example to impose a shape.

**Reply by goerz** [reply]
Yes, of course. You can put in whatever amplitudes you want.

**Comment by goerz** [comment]
More generally, I would also point you to the section on [Arbitrary Generators](https://juliaquantumcontrol.github.io/QuantumPropagators.jl/dev/generators/#Arbitrary-Generators) (again, for the current `master`, the `check_…` routines haven't been released yet).

**Reply by will-atom** [reply]
Following up on this question, how would you suggest specifying a control for a constant amplitude field whose phase is a control parameter? Splitting into real and imaginary parts for this (as in the example) is somewhat inconvenient.

**Reply by goerz** [reply]
Maybe transform to a rotating frame or something where that phase becomes a shift of energy levels? In any case, you'll have to find out some way to write your problem in terms of real controls.

**Comment by will-atom** [comment]
Thanks very much!

**Reply by goerz** [reply]
You should be good to go now if you're running QuantumControl 0.8.0.

If you end up with a nice example notebook for this, I wouldn't mind adding it to the documentation and tests in some way. Let me know.
